apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/functions/CreateSecurityGroupIfNeeded.java

```diff
@@ -17,73 +17,202 @@
+                  nameEquals(regionSecurityGroupNameAndPorts.getName()));
+            logger.debug("<< reused securityGroup(%s)", group.getId());
+            return securityGroupInRegionSecurityGroupFunction
+                  .apply(new SecurityGroupInRegion(group, regionId, allGroups));
+         }
+      }
+   }
+
+   private SecurityGroup createSecurityGroupFrom(final org.jclouds.openstack.neutron.v2.domain.SecurityGroup group,
+         Location location, Set<Integer> ports) {
+      SecurityGroup securityGroup = toSecurityGroup(location).apply(group);
+      logger.debug("<< created securityGroup(%s)", securityGroup);
+
+      SecurityGroupApi securityGroupApi = getNeutronSecurityGroupApi(location.getId());
+      try {
+         for (int inboundPort : ports) {
+            logger.debug(">> authorizing securityGroup(%s) permission to 0.0.0.0/0 on port %d", securityGroup, inboundPort);
+            securityGroupApi.create(
+                  Rule.CreateRule.createBuilder(RuleDirection.INGRESS, RegionAndId.fromSlashEncoded(securityGroup.getId()).getId()).protocol(RuleProtocol.TCP)
+                        .portRangeMin(inboundPort).portRangeMax(inboundPort).remoteIpPrefix("0.0.0.0/0").build());
+            logger.debug("<< authorized securityGroup(%s) permission to 0.0.0.0/0 on port %d", securityGroup, inboundPort);
          }
-         return new SecurityGroupInRegion(api.get().get(securityGroup.getId()), regionId, allGroups);
+         return securityGroup;
       } catch (IllegalStateException e) {
-         logger.trace("<< trying to find securityGroup(%s): %s", regionSecurityGroupNameAndPorts, e.getMessage());
-         SecurityGroup group = find(allGroups, nameEquals(regionSecurityGroupNameAndPorts.getName()));
-         logger.debug("<< reused securityGroup(%s)", group.getId());
-         return new SecurityGroupInRegion(group, regionId, allGroups);
+         logger.trace("<< trying to find securityGroup(%s): %s", group, e.getMessage());
+
+         return securityGroupApi.listSecurityGroups().concat()
+               .filter(new Predicate<org.jclouds.openstack.neutron.v2.domain.SecurityGroup>() {
+                  @Override
+                  public boolean apply(@Nullable org.jclouds.openstack.neutron.v2.domain.SecurityGroup input) {
+                     return input.getName().equals(group.getName());
+                  }
+               }).transform(toSecurityGroup(location)).first().orNull();
       }
    }
 
-   private void authorizeGroupToItselfAndAllIPsToTCPPort(SecurityGroupApi securityGroupApi,
-                                                         SecurityGroup securityGroup, int port) {
+   private SecurityGroupApi getNeutronSecurityGroupApi(String region) {
+      if (neutronContextSupplier == null)
+         return null;
+      return ((ApiContext<NeutronApi>) neutronContextSupplier.get()).getApi().getSecurityGroupApi(region);
+   }
+
+   // TODO remove duplications
```

Comment:
You should be able to remove this when you have the function in a new type using assisted inject.